libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2852,15 +2852,18 @@ string YulUtilFunctions::conversionFunction(Type const& _from, Type const& _to)
 		solAssert(_from.dataStoredIn(DataLocation::CallData), "");
 		solAssert(_to.category() == Type::Category::Array, "");
 
-		ArraySliceType const& fromType = dynamic_cast<ArraySliceType const&>(_from);
-		ArrayType const& targetType = dynamic_cast<ArrayType const&>(_to);
+		auto const& fromType = dynamic_cast<ArraySliceType const&>(_from);
+		auto const& targetType = dynamic_cast<ArrayType const&>(_to);
 
 		solAssert(!fromType.arrayType().baseType()->isDynamicallyEncoded(), "");
 		solAssert(
 			*fromType.arrayType().baseType() == *targetType.baseType(),
 			"Converting arrays of different type is not possible"
 		);
 
+		if (!targetType.dataStoredIn(DataLocation::CallData))
+			return arrayConversionFunction(fromType.arrayType(), targetType);
+
 		string const functionName =
 			"convert_" +
 			_from.identifier() +
```

Comment:
Are there any calldata -> calldata conversions at all?

Reply:
I'm not sure, I originally thought that [calldata_dynamic_forward.sol](https://github.com/ethereum/solidity/blob/develop/test/libsolidity/syntaxTests/array/slice/calldata_dynamic_forward.sol) should be an example of that but, as you explaned, that's an external call and needs a memory argument.

In that case how about passing calldata slice to an internal function that accepts a calldata array?

If that's not possible then this is just a piece of dead code and should actually be removed.

Reply:
conversion to calldata doesn't happen, except `ArraySliceType` -> `ArrayType` which happens here. That is happening when doing index access for calldata array slice, for example in https://github.com/ethereum/solidity/blob/develop/test/libsolidity/semanticTests/array/calldata_slice_access.sol

I would not add `calldata` to `calldata` case in `arrayConversionFunction` as it can be misleading. Regarding test what @cameel suggested, I added it and there is no conversion from `calldata` -> `calldata` at all. Only in case from array slice -> calldata array. Also, conversions in calldata are not possible. For example passing `uint128[]` as an argument to function accepting `uint256[]` is not possible (`TypeChecker` doesn't allow that)